src/graph/immutable_graph.cc

```diff
@@ -4,150 +4,179 @@
  * \brief DGL immutable graph index implementation
  */
 
-#include <string.h>
-#include <sys/types.h>
 #include <dgl/immutable_graph.h>
-
-#ifdef _MSC_VER
-#define _CRT_RAND_S
-#endif
+#include <string.h>
+#include <bitset>
+#include <numeric>
 
 #include "../c_api_common.h"
 
 namespace dgl {
+namespace {
+/*!
+ * \brief A hashmap that maps each ids in the given array to new ids starting from zero.
+ */
+class IdHashMap {
+ public:
+  // Construct the hashmap using the given id arrays.
+  // The id array could contain duplicates.
+  explicit IdHashMap(IdArray ids) {
+    const dgl_id_t* ids_data = static_cast<dgl_id_t*>(ids->data);
+    const int64_t len = ids->shape[0];
+    dgl_id_t newid = 0;
+    for (int64_t i = 0; i < len; ++i) {
+      const dgl_id_t id = ids_data[i];
+      if (!Contains(id)) {
+        oldv2newv_[id] = newid++;
+        filter_.flip(id & kFilterMask);
+      }
+    }
+  }
 
-template<class ForwardIt, class T>
-bool binary_search(ForwardIt first, ForwardIt last, const T& value) {
-  first = std::lower_bound(first, last, value);
-  return (!(first == last) && !(value < *first));
-}
+  // Return true if the given id is contained in this hashmap.
+  bool Contains(dgl_id_t id) const {
+    return filter_.test(id & kFilterMask) && oldv2newv_.count(id);
+  }
+
+  // Return the new id of the given id.
+  dgl_id_t Map(dgl_id_t id) const {
+    return oldv2newv_.at(id);
```

Comment:
this implementation seems to require much more hashtable lookups. it can potentially significantly slow down sampling.

Reply:
https://github.com/dmlc/dgl/blob/a7d0b8a0a416c0a253215e714ce274bb9dd88518/src/graph/immutable_graph.cc#L308-L309

So you think it is better to change it to sth like?
```c++
newv = hashMap.ContainsAndMap(...);
if (IsValid(newv)) {
  // do something.
}
```

Reply:
yes, something like this.

Reply:
it seems it's not used only for node_subgraph and edge_subgraph instead of neighbor sampling. so far, these two operations aren't commonly used. i guess it's fine for now if you think it's not necessary to change it.

Reply:
I could change that. It's easy.

Reply:
Done. Please see whether it looks good to you.